File: 54-spiral-matrix/spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> list = new ArrayList<>();
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0)
            return list;
        int top = 0, left = 0;
        int bottom = matrix.length - 1, right = matrix[0].length - 1;
        while(top <= bottom && left <= right) {
            int i;
            for (i = left; i <= right; i++) {
                list.add(matrix[top][i]);
            }
            top++;
            for (i = top; i <= bottom; i++)
                list.add(matrix[i][right]);
            right--;
            if (top <= bottom) {
                for (i = right; i >= left; i--) {
                    list.add(matrix[bottom][i]);
                }
                bottom--;
            }
            if (left <= right) {
                for (i = bottom; i >= top; i--) {
                    list.add(matrix[i][left]);
                }
                left++;
            }
        }
        return list;
    }
}

File: 54-spiral-matrix/spiral-matrix.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
#include <stdlib.h>

int* spiralOrder(int** matrix, int matrixSize, int* matrixColSize, int* returnSize) {
    int* result = (int*)malloc(matrixSize * (*matrixColSize) * sizeof(int));
    *returnSize = 0;
    
    if (matrixSize == 0 || *matrixColSize == 0) 
        return result;

    int top = 0, left = 0;
    int bottom = matrixSize - 1, right = (*matrixColSize) - 1;

    while (top <= bottom && left <= right) {
        for (int i = left; i <= right; i++) 
            result[(*returnSize)++] = matrix[top][i];
        top++;

        for (int i = top; i <= bottom; i++) 
            result[(*returnSize)++] = matrix[i][right];
        right--;

        if (top <= bottom) {
            for (int i = right; i >= left; i--) 
                result[(*returnSize)++] = matrix[bottom][i];
            bottom--;
        }

        if (left <= right) {
            for (int i = bottom; i >= top; i--) 
                result[(*returnSize)++] = matrix[i][left];
            left++;
        }
    }

    return result;
}
